README.md
# PortaDOOM #

A self-contained, portable [DOOM](https://en.wikipedia.org/wiki/Doom_(1993_video_game)) launcher and [WAD](https://en.wikipedia.org/wiki/Doom_WAD) collection presented as an early '90s MS-DOS [diskzine](https://en.wikipedia.org/wiki/Disk_magazine).

![PortaDOOM Screenshot](screenshot.png)

It's intended as a single download that people can unzip and play on any computer, targeted at people who have either never played DOOM or anything beyond the official iD canon and the thought of downloading, configuring and launching several different engines with all kinds of weird WAD files is putting them off trying stuff out.

PortaDOOM is about having a single executable where you press a few keys and hey-presto! you're playing DOOM WADs in the correct engine with the correct settings.

The viewer is written in [QB64](http://www.qb64.net/); a modern remake of Microsoft's classic [QuickBASIC](https://en.wikipedia.org/wiki/QuickBASIC) for that DOS look and feel, but you do not need to be able to code to contribute: all the diskzine pages are just text-files that anybody can create and edit.

You can contribute the following ways:
*   Write pages yourself using [this handy guide](DOSmag/README.md#how-to-write-pages)
*   File GitHub Issues for suggestions and fixes
*   E-mail me: <kroc@camendesign.com>

Have fun and DOOM on!

CHANGELOG.TXT
+ = New feature / addition
* = Fix / Change
- = Removal

02-MAY-19:
* Updated GZDoom to 4.1.0

14-APR-19:
* Updated GZDoom to 4.0.0

09-MAR-19:
* Updated Crispy-Doom to v5.5.1
* Updated DOOM-Retro to v2.8.1

22-JAN-19: v2018.12.31/3
* Updated GZDoom to v.3.7.2
* Move stable release back to GZDoom v3.7 (initial release had bugs)

03-JAN-19: v2018.12.31/2
* Updated GZDoom to v3.7.1 (overall)
* Went back to GZDoom 3.6.0 for Cacowards 2018 due to ongoing reliability
  problems with GZDoom 3.7.x

01-DEC-19: v2018.12.31/1
* Fix to Launcher.exe hanging (OpenGL detection issue)

31-DEC-18: v2018.12.31/0
+ Released Cacowards 2018!

22-DEC-18:
* Updated Adventures of Square, added launcher IWAD detection
+ Added Adventures of Square to default config files
- CHEX support appears to have disappeared in Crispy-Doom

29-DEC-18:
+ Added GZDoom 3.7.0

18-DEC-18:
* Crispy-doom updated to v5.4

17-NOV-18:
* Updated DOOM Retro to v2.7.5

03-NOV-18:
* Updated DOOM Retro to v2.7.4
  (this release allows binding the F12 key for screenshots)

11-OCT-18:
* Updated GZDoom to v3.6.0

11-SEP-18:
* Updated DOOM Retro to v2.7.3
- Removed GZDoom v0.9 (I don't think anything will require it)

29-AUG-18:
* Updated UPX to 3.9.5

28-AUG-18:
* Update DOOM Retro to v2.7.2
* Update GZDoom to v3.5.1

13-AUG-18:
* Update Crispy-Doom to v2.3.0

06-AUG-18:
* Update DOOM Retro to v2.7.1

31-JUL-18:
+ Added GZDoom v3.5.0
+ Set online stats collection to off by default for GZDoom engines

30-JUL-18:
+ Added Innocence X / X2

27-JUN-18:
* Switched to a Git Submodule for QB64

20-JUN-18:
+ Added GZDoom v3.4.1 
* Updated Crispy Doom to 5.2

25-APR-18:
* Updated GZDoom to v3.3.2

20-APR-18:
* Updated GZDoom to v3.3.1
* DOOM Retro updated to v2.6.9
* Improvements to "build.bat" (zip command is more modular)

08-APR-18:
* Fixed typo in doom.bat that was preventing anything from working!
  How did this slip through, when I tested each release??

26-MAR-18:
+ Added GZDoom v3.3
+ Added GZDoom-dev

19-MAR-18:
+ PSX DOOM now resets weapons between 'episodes'
+ Added bloom effect to Doomed Space Wars ("spacwars.wad")
* A new launcher is being built to solve the issue of managing mod selection &
  compatibility. A UI interface is being built up which includes functions for
  formatted string printing. DOSmag has been re-factored to use this, greatly
  reducing the complexity of printing the formatted lines there
* "The Given" updated to final release. Adds a Low-Detail mode.
  This needs to be added to the page I think
* mods have been moved outside of wads folder, this will help in the future
  with the custom launcher
* Numerous pages have minor formatting corrections to account for the different
  method of processing

23-FEB-18:
* Removed DUMP3, it's very buggy and not worth the effort right now

21-JAN-18:
* Upgraded DOOM Retro to 2.6.4; this fixes the config-file issue so the engine
  can be included proper now
* DUMP3 now uses GZDoom v2.2 due to compatibility bug with later versions

17-JAN-18:
* QB64 upgraded to v1.2 -- improved DOSmag compilation
* Re-instated blinking effect in DOSmag whilst keeping 30 rows (640x480)
* Slight re-organisation of DOSmag code

16-JAN-18:
+ Added `/EXEC` parameter in "play.bat" to pass through to "doom.bat"
+ Enabled bloom effect for "Shadows of the Nighmare Realm", "Void & Rainbow",
  "dead.air", "dead.wire" and "ALIENS (TC) 2017"
* Fixed "play.bat" parsing of multiple `/REQ` tags (uses ":" now),
  using "," or ";" are treated as parameter separators in batch!
* Changed "Shadow of The Wool Ball" to use ZDoom (instead of GZDoom)

15-JAN-18:
+ "doom.bat" now offers to generate "DOOM64.WAD" from an N64 ROM
* Fixed FreeDOOM substitution in "doom.bat"

14-JAN-18:
+ Upgraded Crispy Doom to v5.1

08-JAN-18:
+ Added "Dark Tide" WAD page
+ Added "DOOM II: Hellscape" WAD page

06-JAN-18:
+ Added "Water Spirit" WAD page

05-JAN-18:
+ Added "Disjunction" WAD page
+ Added "Waterlab GZD" WAD page
+ Added "Hexen - Curse of the Lost Gods" WAD page
* Updated GZDoom to 3.2.5

04-JAN-18:
* Engine default configuration files re-done with a new script to automatically
  modify the multitude of config files with our desired settings

30-DEC-17:
* Upgraded Chocolate Doom to 3.0
* Changed "doom.bat" to separate different GZDoom versions' config files

28-DEC-17:
+ Added STRIFE Steam / GOG detection and shareware substitution
* "play.bat" changed to use "," as `/REQ` tag separator and new tags added

24-DEC-17:
* Fixed sound in DOOM 64 EX not working.
  Binaries were updated to fix mouse issue in Windows 10

23-DEC-17:
+ Added Crispy Doom engine to "doom.bat" -- vanilla, no-limits
+ Added DOOM Retro engine to "doom.bat"  -- boom

22-DEC-17:
+ Added "Return to Hadron Episode 2" WAD page
+ Added "TNT: Revilution" WAD page

21-DEC-17:
+ "brightmaps.pk3" & "lights.pk3" (GZDoom), or "skulltag_actors.pk3" &
  "skulltag_data.pk3" (Zandronum) inserted automatically, before the PWAD,
  allowing these to be correctly replaced by the PWAD
- Removed the insertion of "brightmaps.pk3" / "lights.pk3" from play.bat,
  as these would come _after_ the PWAD, causing potential problems

19-DEC-17:
+ Added GZDoom versions 0.9 to 2.1!
+ Added a build target for just the launcher ("doom.bat" / "play.bat")
+ Added "Rise of the Wool Ball" WAD page
- Removed GZDoom v3.1 due to security issue in version 3.0-3.2.3

17-DEC-17:
+ Added "Stardate 20x7" WAD page
+ Added "Void and Rainbow" WAD page
* Upgraded GZDoom to 3.2.4

15-DEC-17:
+ Added "Legacy of Heroes" WAD page
+ Added "Saturnine Chapel" WAD page

14-DEC-17:
+ Added "3 Heures d'agonie" series by the French Doom Community
+ Added "dead.air" WAD page
+ Added "Brigandine" WAD page
+ Added "Counterattack" WAD page
+ Added "The Given" WAD page

13-DEC-17:
* Created a Doom the Way id Did Series page set
* Changed a number of links from `*[N] Name*` to `[N: Name]`

11-DEC-17:
+ Added Cacowards 2017 build target (incomplete)
+ Added ZDoom v2.8.1 back to doom.bat, required for lilith.pk3

consts.bas
'DOSmag : a DOS-like portable front-end for hyperlinked textual content.
'         copyright (C) Kroc Camen, 2016-2017; MIT license (see LICENSE.TXT)
'=============================================================================
'WARNING: THIS IS A QB64.NET SOURCE FILE, ENCODED AS ANSI CODE-PAGE 437
'         (SOMETIMES REFERRED TO AS "DOS" OR "OEM-US" ENCODING).
'         DO NOT OPEN AND SAVE THIS FILE AS UNICODE / UTF-8!
'=============================================================================

CONST TRUE = -1
CONST FALSE = 0

'------------------------------------------------------------------------------

'the default size (in char cols/rows) of the text-mode screen (`SCREEN 0`)
CONST SCREEN_WIDTH = 80
CONST SCREEN_HEIGHT = 25

'------------------------------------------------------------------------------

CONST BLACK = 0 '..COLOR code for black
CONST BLUE = 1 '...COLOR code for dark blue
CONST GREEN = 2 '..COLOR code for dark green
CONST CYAN = 3 '...COLOR code for dark-cyan
CONST RED = 4 '....COLOR code for dark-red
CONST PURPLE = 5 '.COLOR code for purple
CONST ORANGE = 6 '.COLOR code for brown / orange
CONST LTGREY = 7 '.COLOR code for "white" (light grey)
CONST DKGREY = 8 '.COLOR code for "bright black" (dark grey), foreground ONLY
CONST LTBLUE = 9 '.COLOR code for "bright blue", foreground ONLY
CONST LIME = 10 '..COLOR code for "bright green", foreground ONLY
CONST AQUA = 11 '..COLOR code for "bright cyan", foreground ONLY
CONST ROSE = 12 '..COLOR code for "bright red" (peach), foreground ONLY
CONST PINK = 13 '..COLOR code for "bright purple" (fuscia), foreground ONLY
CONST YELLOW = 14 'COLOR code for "bright orange" (yellow), foreground ONLY
CONST WHITE = 15 '.COLOR code for "bright white", foreground ONLY

CONST BLINK = 16 '.add this to any colour to make it blink on & off

'------------------------------------------------------------------------------

CONST ASC_NULL = 0

CONST ASC_BKSP = 8
CONST ASC_TAB = 9

CONST ASC_ESC = 27 '      escape key

CONST ASC_SPC = 32 '      space
CONST ASC_EXCL = 33 '     !
CONST ASC_SMARK = 34 '    "
CONST ASC_HASH = 35 '     #
CONST ASC_APOS = 39 '     '
CONST ASC_LPAREN = 40 '   (
CONST ASC_RPAREN = 41 '   )
CONST ASC_ASTERISK = 42 ' *
CONST ASC_COMMA = 44 '    ,
CONST ASC_DASH = 45 '     - (hyphen / minus)
CONST ASC_PERIOD = 46 '   . (fullstop)
CONST ASC_FSLASH = 47 '   / (forward-slash)
CONST ASC_COLON = 58 '    :
CONST ASC_SEMICOLON = 59 ';
CONST ASC_EQUALS = 61 '   =
CONST ASC_QMARK = 63 '    ?
CONST ASC_AT = 64 '       @

CONST ASC_A = 65
CONST ASC_B = 66
CONST ASC_C = 67
CONST ASC_D = 68
CONST ASC_E = 69
CONST ASC_F = 70
CONST ASC_G = 71
CONST ASC_H = 72
CONST ASC_I = 73
CONST ASC_J = 74
CONST ASC_K = 75
CONST ASC_L = 76
CONST ASC_M = 77
CONST ASC_N = 78
CONST ASC_O = 79
CONST ASC_P = 80
CONST ASC_Q = 81
CONST ASC_R = 82
CONST ASC_S = 83
CONST ASC_T = 84
CONST ASC_U = 85
CONST ASC_V = 86
CONST ASC_W = 87
CONST ASC_X = 88
CONST ASC_Y = 89
CONST ASC_Z = 90

CONST ASC_LSQB = 91 '   [
CONST ASC_BSLASH = 92 ' \
CONST ASC_RSQB = 93 '   ]
CONST ASC_CARET = 94 '  ^
CONST ASC_USCORE = 95 ' _
CONST ASC_BAR = 124 '   |

'ASCII symbols specific to the DOS codepage:
'-----------------------------------------------------------------------------

CONST ASC_DIAMOND = 4

CONST ASC_ARR_UP = 24 '.......up-arrow
CONST ASC_ARR_DN = 25 '.......down-arrow
CONST ASC_ARR_LT = 27 '.......left-arrow
CONST ASC_ARR_RT = 26 '.......right-arrow

CONST ASC_LGLLMT = 174 '......left guillemet, i.e. double angle-bracket

CONST ASC_BOX_H = 196 '.......horizontal line
CONST ASC_BOX_V = 179 '.......vertical line
CONST ASC_BOX_TL = 218 '......top-left corner
CONST ASC_BOX_TR = 191 '......top-right corner
CONST ASC_BOX_BL = 192 '......bottom-left corner
CONST ASC_BOX_BR = 217 '......bottom-right corner

CONST ASC_BOX_DBL_H = 205 '...horizontal double-line

CONST ASC_BOX_BL_DBL_B = 212 'bottom-left corner, with double-line on bottom
CONST ASC_BOX_BR_DBL_B = 190 'bottom-right corner, double-line on bottom


'when the result of INKEY$ begins with a NULL,
'then the second byte is the following:
'-----------------------------------------------------------------------------
CONST ASC_F1 = 59
CONST ASC_F2 = 60
CONST ASC_F3 = 61
CONST ASC_F4 = 62
CONST ASC_F5 = 63
CONST ASC_F6 = 64
CONST ASC_F7 = 65
CONST ASC_F8 = 66
CONST ASC_F9 = 67
CONST ASC_F10 = 68
CONST ASC_F11 = 133
CONST ASC_F12 = 134

CONST ASC_UP = 72
CONST ASC_DOWN = 80
CONST ASC_LEFT = 75
CONST ASC_RIGHT = 77

CONST ASC_HOME = 71
CONST ASC_PGUP = 73
CONST ASC_PGDN = 81
CONST ASC_END = 79
